Mathlib/Analysis/Complex/Periodic.lean

```diff
@@ -0,0 +1,214 @@
+
+open scoped Real Topology
+
+noncomputable section
+
+local notation "I∞" => comap im atTop
+
+section q_param
+
+variable (h : ℝ)
+
+/-- Parameter for q-expansions, `q_param h z = exp (2 * π * I * z / h)` -/
+def q_param (z : ℂ) : ℂ := exp (2 * π * I * z / h)
+
+/-- One-sided inverse of `q_param h`. -/
+def inv_q_param (q : ℂ) : ℂ := h / (2 * π * I) * log q
+
+theorem abs_q_param (z : ℂ) : abs (q_param h z) = Real.exp (-2 * π * im z / h) := by
+  simp only [q_param, abs_exp, div_ofReal_re, mul_re, re_ofNat, ofReal_re, im_ofNat, ofReal_im,
+    mul_zero, sub_zero, I_re, mul_im, zero_mul, add_zero, I_im, mul_one, sub_self, zero_sub,
+    neg_mul]
+
+theorem im_inv_q_param (q : ℂ) : im (inv_q_param h q) = -h / (2 * π) * Real.log (abs q) := by
+  simp only [inv_q_param, ← div_div, div_I, neg_mul, neg_im, mul_im, mul_re, div_ofReal_re,
+    div_ofNat_re, ofReal_re, I_re, mul_zero, div_ofReal_im, div_ofNat_im, ofReal_im, zero_div, I_im,
+    mul_one, sub_self, zero_mul, add_zero, log_re, zero_add, neg_div]
+
+variable {h} -- next few theorems all assume h ≠ 0 or 0 < h
+
+theorem q_param_right_inv (hh : h ≠ 0) {e : ℂ} (hq : e ≠ 0) :
+    q_param h (inv_q_param h e) = e := by
+  simp only [q_param, inv_q_param, ← mul_assoc, mul_div_cancel₀ _ two_pi_I_ne_zero,
+    mul_div_cancel_left₀ _ (ofReal_ne_zero.mpr hh), exp_log hq]
+
+theorem q_param_left_inv_mod_period (hh : h ≠ 0) (z : ℂ) :
+    ∃ m : ℤ, inv_q_param h (q_param h z) = z + m * h := by
+  dsimp only [q_param, inv_q_param]
+  obtain ⟨m, hm⟩ := log_exp_exists (2 * ↑π * I * z / ↑h)
+  use m
+  rw [hm, mul_div_assoc, mul_comm (m : ℂ), ← mul_add, ← mul_assoc,
+    div_mul_cancel₀ _ two_pi_I_ne_zero, mul_add, mul_div_cancel₀ _ (mod_cast hh), mul_comm]
+
+theorem abs_q_param_lt_iff (hh : 0 < h) (A : ℝ) (z : ℂ) :
+    abs (q_param h z) < Real.exp (-2 * π * A / h) ↔ A < im z := by
+  rw [abs_q_param, Real.exp_lt_exp, div_lt_div_right hh, mul_lt_mul_left_of_neg]
+  simpa using Real.pi_pos
+
+theorem q_param_tendsto (hh : 0 < h) : Tendsto (q_param h) I∞ (𝓝[≠] 0) := by
+  refine tendsto_nhdsWithin_of_tendsto_nhds_of_eventually_within _ ?_
+    (.of_forall fun q ↦ exp_ne_zero _)
+  rw [tendsto_zero_iff_norm_tendsto_zero]
+  simp only [norm_eq_abs, abs_q_param]
+  apply (tendsto_comap'_iff (m := fun y ↦ Real.exp (-2 * π * y / h)) (range_im ▸ univ_mem)).mpr
+  refine Real.tendsto_exp_atBot.comp (.atBot_div_const hh (tendsto_id.const_mul_atTop_of_neg ?_))
+  simpa using Real.pi_pos
+
+theorem inv_q_param_tendsto (hh : 0 < h) : Tendsto (inv_q_param h) (𝓝[≠] 0) I∞ := by
+  simp only [tendsto_comap_iff, comp_def, im_inv_q_param]
+  apply Tendsto.const_mul_atBot_of_neg (div_neg_of_neg_of_pos (neg_lt_zero.mpr hh) (by positivity))
+  exact Real.tendsto_log_nhdsWithin_zero_right.comp tendsto_norm_nhdsWithin_zero
+
+end q_param
+
+section PeriodicOnℂ
+
+variable (h : ℝ) (f : ℂ → ℂ)
+
+/-- The function `q ↦ f (inv_q_param h q)`, extended by a non-canonical choice of limit at 0. -/
+def cuspFcn : ℂ → ℂ :=
```

Comment:
Does "fcn" mean something?

Reply:
It was intended as shorthand for "function". I'd be happy to rename if there's something else which fits better.

Reply:
How about ... "function"?

Reply:
Sorry, my last post was very unclear; I meant that there might be better alternatives than "cuspFunction" or abbreviations thereof.

Reply:
:bulb: :+1: